floodlight/core/events.py

```diff
@@ -61,3 +95,148 @@ def custom(self):
+            col
+            for col in self.protected
+            if not self.column_values_in_range(col, protected_columns)
+        ]
+        if not incorrect_columns:
+            incorrect_columns = None
+
+        return incorrect_columns
+
+    def column_values_in_range(self, col: str, definitions: Dict[str, Dict]) -> bool:
+        """Perform the checks for value range for a single column of the inner event
+        DataFrame using the specifications from floodlight.core.definitions.
+
+        Parameters
+        ----------
+        col: str
+            Column of the inner event DataFrame to be checked
+        definitions: Dict
+            Dictionary for each column name to be checked. Each entry is a Dictionary
+            containing definitions for allowed value ranges of the respective
+            column. Information about value_range is given as a List of the form
+            [min_value, max_value].
+
+        Returns
+        -------
+        bool
+            True if the checks for value range pass and False otherwise
+        """
+        # skip if value range is not defined
+        if definitions[col]["value_range"] is None:
+            return True
+
+        # skip values that are None or NaN
+        check_col = self.events[col].dropna()
+
+        # check value range for remaining values
+        if not (definitions[col]["value_range"][0] <= check_col).all():
+            return False
+        if not (check_col <= definitions[col]["value_range"][1]).all():
+            return False
+
+        # all checks passed
+        return True
+
+    def add_frameclock(self, framerate: int):
+        """Add the column "frameclock", computed as the rounded multiplication of
+        gameclock and framerate, to the inner events DataFrame.
+
+        Parameters
+        ----------
+        framerate: int
+            Temporal resolution of data in frames per second/Hertz.
+        """
+        frameclock = np.full((len(self.events)), -1, dtype=int)
+        frameclock[:] = np.floor(self.events["gameclock"].values * framerate)
+        self.events["frameclock"] = frameclock
+
+    def find(
+        self, conditions: Tuple[str, Any] or List[Tuple[str, Any]]
+    ) -> pd.DataFrame:
+        """Returns a DataFrame containing all entries from the inner events DataFrame
+         that satisfy all given conditions.
+
+        Parameters
+        ----------
+        conditions: Tuple or List of Tuples
+            Conditions used to filter the columns. Can be either a single Tuple or a
+            List of Tuples. If given as a List, all conditions are enforced. Conditions
+            can be of the form (column, value) or of the form (column, (min, max)). In
+            case of the former, the specified column of the inner events DataFrame is
+            searched for entries equal to value. In case of the latter, the specified
+            column is searched for values in the range (min, max).
+            entries.
```

Comment:
```suggestion
            A single or a list of conditions used for filtering. Each condition should 
            follow the form ``(column, value)``. Here, ``value`` can be either a variable 
            (to filter for an exact value), or a tuple ``value = (min, max)`` that 
            specifies a minimum and maximum value (to filter for a value range).
            
            For example, to filter all events that have the ``eID`` of ``"Pass"`` and 
            that happened within the first 1000 seconds of the segment,
            conditions should look like:
            ``conditions = [("eID", 1), ("gameclock", (0, 1000))]
            
            You may also pass a condition where ``value = None``. In this case...
```

Reply:
I found your description of ``conditions`` slightly confusing, and added an alternative formulation. Maybe try to first explain the syntax, then the semantics behind it. Also, constructions involving "the former" or "the latter" quickly lead to many lines of explanation that are harder to understand. Sometimes, an example is also much easier to understand.

Reply:
The use of double backticks for code-related wording (that includes variables) will also really help readability after the docstring is rendered

Reply:
Same as above, it might be easier to treat the possibility of value being None similar to value being a single variable: 

        Parameters
        ----------
        conditions: Tuple or List of Tuples
            A single or a list of conditions used for filtering. Each condition should
            follow the form ``(column, value)``. If ``value`` is given as a variable
            (can also be None), it is used to filter for an exact value. If given as a 
            tuple ``value = (min, max)`` that specifies a minimum and maximum value, it
            is filtered for a value range.

            For example, to filter all events that have the ``eID`` of ``"Pass"`` and
            that happened within the first 1000 seconds of the segment, conditions
            should look like:
            ``conditions = [("eID", 1), ("gameclock", (0, 1000))]